New SnowflakeQueryFromFile Task and SQLServerExecute Fix

kvnkho opened the pull request:
<!-- Thanks for contributing to Prefect Core! 🎉-->

## Summary

This creates a new task for SnowflakeQueryFromFile. Some users have been looping through .sql files with the ShellTask so this should help them.

The second fix is fixing the MS SQL Server execute Task. This was broken because the library can not take in kwargs. This fix is too hard to test I think.

## Checklist
<!-- PRs will not be reviewed unless these boxes are checked -->

**Edit** - I edited the SnowflakeQueryFromFile to SnowflakeQueriesFromFile. It now takes in multiple SQL queries and returns a list of results of the queries.

This PR:

- [x] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

kvnkho commented:
Don't approve and merge this yet. I got feedback that this does not work in handling multiple queries in one go. It might be a limitation of the Snowflake connector, but I have to check.

kvnkho commented:
This is done.

madkinsz commented on src/prefect/tasks/snowflake/snowflake.py:
```diff
@@ -114,3 +115,117 @@ def run(self, query: str = None, data: tuple = None, autocommit: bool = None):
+        self.private_key = private_key
+        super().__init__(**kwargs)
+
+    @defaults_from_attrs("file_path", "autocommit")
+    def run(self, file_path: str = None, autocommit: bool = None):
```
This won't allow users to pass the password as a Secret. All of the options should be available at runtime.

madkinsz commented on tests/tasks/snowflake/test_snowflake.py:
```diff
@@ -76,3 +76,46 @@ def test_execute_fetchall(self, monkeypatch):
+        snowflake_module_connect_method = MagicMock()
+        connection = MagicMock(spec=sf.SnowflakeConnection)
+
+        # link all the mocks together appropriately
+        snowflake_module_connect_method.return_value = connection
```
You can just define `snowflake_module_connect_method = MagicMock(return_value=connection)`

madkinsz commented on tests/tasks/snowflake/test_snowflake.py:
```diff
@@ -76,3 +76,46 @@ def test_execute_fetchall(self, monkeypatch):
         assert output == ["TESTDB"]
+
+
+class TestSnowflakeQueriesFromFile:
```
There should be a test that covers the functionality i.e. that the correct calls are made to the connection and that mocked return values are handled correctly

madkinsz reviewed (changes requested):
What is the `SQLServerExecute` fix?

kvnkho commented:
> What is the `SQLServerExecute` fix?

`cursor.execute(query=query, vars=data)` breaks because execute cannot take kwargs.

zangell44 commented on src/prefect/tasks/mysql/mysql.py:
```diff
@@ -46,32 +46,58 @@ def __init__(
-            db=self.db_name,
-            charset=self.charset,
-            port=self.port,
+            host=(host or self.host),
```
the decorator on this line takes care of this logic for you https://github.com/PrefectHQ/prefect/pull/4363/files#diff-dcbaaa9d19c03a04f422b74f5edbc0a28717268dd9b4492320337d74b082bafcR49

zangell44 commented on src/prefect/tasks/mysql/mysql.py:
```diff
@@ -205,12 +258,12 @@ def run(
-            db=self.db_name,
-            charset=self.charset,
-            port=self.port,
+            host=(host or self.host),
```
again `defaults_from_attrs` decorator will take care of this for you

zangell44 commented on src/prefect/tasks/snowflake/snowflake.py:
```diff
@@ -47,51 +48,91 @@ def __init__(
-            "role": self.role,
-            "warehouse": self.warehouse,
-            "autocommit": self.autocommit,
+            "account": (account or self.account),
```
again, no need for this logic

zangell44 commented on src/prefect/tasks/snowflake/snowflake.py:
```diff
@@ -114,3 +155,158 @@ def run(self, query: str = None, data: tuple = None, autocommit: bool = None):
+
+        # build the connection parameter dictionary
+        # we will remove `None` values next
+        connect_params = {
```
I bet you can guess what I'll say by now 😄

zangell44 commented on src/prefect/tasks/snowflake/snowflake.py:
```diff
@@ -114,3 +155,158 @@ def run(self, query: str = None, data: tuple = None, autocommit: bool = None):
+                for cursor in cursor_list:
+                    result.append(cursor.fetchall())
+                    # return fetch for each cursor
+            conn.close()
```
for here and below, doesn't the context manager take care of this for us?

zangell44 commented on src/prefect/tasks/snowflake/snowflake.py:
```diff
@@ -114,3 +155,158 @@ def run(self, query: str = None, data: tuple = None, autocommit: bool = None):
+        }
+
+        # connect to database, open cursor
+        conn = sf.connect(**connect_params)
```
do we need to be careful about managing the connection here? if we hit an exception loading the query from file, is the connection/cursor properly closed?

kvnkho replied:
Nice catch, we should read the file inside the try block. Will update.

zangell44 commented on tests/tasks/snowflake/test_snowflake.py:
```diff
@@ -50,29 +91,121 @@ def test_execute_fetchall(self, monkeypatch):
+        # The result is a list because multiple queries are executed
+        assert output == ["TESTDB"]
+
+    def test_required_parameters(self):
```
I think we want `SnowflakeQueriesFromFile` instead of `SnowflakeQuery` here right?

zangell44 commented on tests/tasks/snowflake/test_snowflake.py:
```diff
@@ -50,29 +91,121 @@ def test_execute_fetchall(self, monkeypatch):
+            account="test", user="test", password="test", file_path=sql_file
+        )
+
+        # The result is a list because multiple queries are executed
```
maybe also assert the connection is established with expected parameters here?

kvnkho replied:
good idea! will do

zangell44 commented on tests/tasks/snowflake/test_snowflake.py:
```diff
@@ -50,29 +91,121 @@ def test_execute_fetchall(self, monkeypatch):
+
+    def test_execute_fetchall(self, monkeypatch, tmpdir, sql_file):
+        """
+        Tests that the SnowflakeQueryFromTime Task calls the fetchall method on the
```
```suggestion
        Tests that the SnowflakeQueriesFromFile Task calls the fetchall method on the
```

zangell44 commented on src/prefect/tasks/snowflake/snowflake.py:
```diff
@@ -47,51 +48,91 @@ def __init__(
 
-    @defaults_from_attrs("query", "data", "autocommit")
-    def run(self, query: str = None, data: tuple = None, autocommit: bool = None):
+    @defaults_from_attrs(
```
should we also include `warehouse` here? (it might not make sense, I'm not very familiar with Snowflake)

zangell44 commented on src/prefect/tasks/snowflake/snowflake.py:
```diff
@@ -114,3 +155,158 @@ def run(self, query: str = None, data: tuple = None, autocommit: bool = None):
+        "file_path",
+        "autocommit",
+    )
+    @defaults_from_attrs("file_path", "autocommit")
```
```suggestion
```

zangell44 commented on src/prefect/tasks/snowflake/snowflake.py:
```diff
@@ -114,3 +155,158 @@ def run(self, query: str = None, data: tuple = None, autocommit: bool = None):
+        self.private_key = private_key
+        super().__init__(**kwargs)
+
+    @defaults_from_attrs(
```
same question for this one, should `warehouse` be included here?

kvnkho replied:
Yes you're right.

kvnkho commented:
I believe this is done now.

zangell44 reviewed (dismissed):
lgtm

kvnkho commented:
I updated this. Couple of changes.

1. I wrote a test for the SqlServerExecute fix. This test makes sure that kwargs are not passed to the `.execute()` method because that throws errors (seen in the linked issue)
2. I updated SnowflakeQueriesFromFile to match the latest updates to SnowflakeQuery from master. Mainly, it takes in a cursor_type that can change the return.
3. I removed MySQL changes from this PR from last time because there were too many merge conflicts with the latest master. It doesn't belong here anyway.

madkinsz commented:
I'm a bit confused about the docs failure, can you push a commit to retrigger? `git commit --allow-empty -m "Force docs rebuild"`
